### PR TITLE
Introduction of Oracles

### PR DESCRIPTION
**Pull request by goran-ethernal**
# Description

This PR introduces a new separation of logic, where every external part is now considered as `Oracle`:

```go
// Oracle represents a feature that can provide and verify system transactions
type Oracle interface {
	// Close closes the oracle
	Close()
	// PostBlock posts a block to the oracle
	PostBlock(postBlockReq *polytypes.PostBlockRequest) error
	// PostEpoch posts an epoch to the oracle
	PostEpoch(postEpochReq *polytypes.PostEpochRequest) error
	// GetTransactions returns the system transactions
	GetTransactions(blockInfo NewBlockInfo) ([]*types.Transaction, error)
	// VerifyTransaction verifies system transactions
	VerifyTransactions(blockInfo NewBlockInfo, txs []*types.Transaction) error
}
```
Oracles have two main tasks, to provide block proposer and active validators that verify the new proposal, the system (state) transactions that need to be added to the block. Each oracle knows on which block it needs to include his transactions (`GetTransactions` function), and also knows how to verify the transactions that are of its concern (`VerifyTransactions` function).

For example, `bridge` oracle, knows when to add `commitBatch` and `commitValidatorSet` transactions, and also knows how to verify them.

There are oracles which do not send transactions, but need to process block and epoch data through `PostBlock` and `PostEpoch` functions, like `governanceManager` or `stakeManager`.

Oracles on blade are:
- `bridge` - creates and validates `commitBatch` and `commitValidatorSet` transactions.
- `epochManager` - creates and validates epoch closing transactions `commitEpoch` and `distributeRewards`.
- `governanceManager` - reads block and epoch data for governance events.
- `stakeManager` - reads block and epoch data for stake change events.
- `proposerCalculator` - reads block data to update the proposer calculation.

### Main changes

- Added a new package called `oracle` that contains `Oracle` interface.
- introduced a new feature called `epochManager` that is one of the oracles tasked with creating and verifying epoch closing transactions (`commitEpoch` and `distributeRewards`).
- `consensus_runtime` now has `Oracles` struct that is a collection of registered oracles.
- Each oracle now handles and verifies its transactions, so that logic is moved from `consensus_runtime` and `fsm` to its corresponding oracles.
- `fsm` will use oracles to write system (state) transactions and verify them. It doesn't need that logic anymore.
- updated Unit Tests accordingly.

### Mockery

This PR introduces `mockery` in blade. It is a tool that auto-generates mocks based on provided configuration in the `.mockery.yml` file.
Mocks are generated into new files in specified packages, and their names are `nameoftheinterface.mock.generated.go`.

To run `mocker` to generate mocks, execute this command in terminal when in `blade` folder:
```
make generate-mocks
```

Since these files are generated by the `mockery` tool, the reviewers don't have to review them.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### CODE REVIEW

File: consensus/polybft/bridge/bridge.go [hunk 1]
```diff
@@ -168,3 +193,253 @@ func (b *bridge) BridgeBatch(pendingBlockNumber uint64) ([]*BridgeBatchSigned, e
...
+		}
+
+		sig := txData[:helpers.AbiMethodIDLength]
+
+		if bytes.Equal(sig, commitBatchFn.Sig()) {
+			if !blockInfo.IsEndOfSprint {
+				return errBridgeBatchTxInNonSprintBlock
+			}
+
+			if bridgeBatchTxExists {
```
Comment (oliverbundalo) [inline consensus/polybft/bridge/bridge.go:263]: Whether this should be an error? We can have multiple batches due to multiple bridges?

Comment (goran-ethernal) [reply consensus/polybft/bridge/bridge.go:263]: We are aware of this. It will be fixed through another PR, when we write tests for multiple bridges.

**Review by oliverbundalo — APPROVED** [review]
LGTM, maybe to update the PR description since ProposerCalculator is oracle as well.

**Review by dusannosovic-ethernal — APPROVED** [review]

File: consensus/polybft/oracle/oracle.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		g.Go(func() error {
+			return oracle.VerifyTransactions(blockInfo, txs)
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (Stefan-Ethernal) [inline consensus/polybft/oracle/oracle.go]: 😎

```suggestion
	return g.Wait()
```

File: consensus/polybft/oracle/oracle.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+// Oracle represents a feature that can provide and verify system transactions
+type Oracle interface {
+	// Close closes the oracle
+	Close()
+	// PostBlock posts a block to the oracle
+	PostBlock(postBlockReq *PostBlockRequest) error
+	// PostEpoch posts an epoch to the oracle
+	PostEpoch(postEpochReq *PostEpochRequest) error
```
Comment (Stefan-Ethernal) [inline consensus/polybft/oracle/oracle.go:49]: Just a thought:

Would it make sense to create a separate interface called e.g. `ReadOnlyOracle` or `SubscribedOracle` etc. that would contain only these two functions.

The thing is, you are currently forcing the `ProposerCalculator` to implement the `Close`, `GetTransactions` and `VerifyTransactions`, although it does not need them, so a better interface seggregation would come handy IMO (although you would need to check the interface type when you want to call a specific `Oracle`).

**Review by Stefan-Ethernal — APPROVED** [review]
I pretty much like the idea.

File: consensus/polybft/oracle/oracle.go (hunk 2 above)
Comment (goran-ethernal) [reply consensus/polybft/oracle/oracle.go]: Fix: https://github.com/Ethernal-Tech/blade/pull/410/commits/b48c515eef226d5465976a54583f7dd83ecfb8b9

File: consensus/polybft/oracle/oracle.go (hunk 3 above)
Comment (goran-ethernal) [reply consensus/polybft/oracle/oracle.go:49]: Fix: https://github.com/Ethernal-Tech/blade/pull/410/commits/b48c515eef226d5465976a54583f7dd83ecfb8b9

**Pull request merged**